src/components/Table/GridTable.tsx

```diff
@@ -297,18 +297,23 @@ export function GridTable<R extends Kinded, S = {}, X extends Only<GridTableXss,
   const virtuosoRef = useRef<VirtuosoHandle | null>(null);
 
   const [sortState, setSortKey] = useSortState<R, S>(columns, sorting);
+  // Disclaimer that technically even though this is a useMemo, sortRows is mutating `rows` directly
   const maybeSorted = useMemo(() => {
     if (sorting?.on === "client" && sortState) {
       // If using client-side sort, the sortState use S = number
-      return sortRows(columns, rows, sortState as any as SortState<number>);
+      sortRows(columns, rows, sortState as any as SortState<number>);
+      return rows;
     }
     return rows;
   }, [columns, rows, sorting, sortState]);
   const noData = !rows.some((row) => row.kind !== "header");
 
-  // For each row, keep it's ReactElement (React.memo'd) + its filter values, so we can re-filter w/o re-evaluating this useMemo.
-  const allRows: RowTuple<R>[] = useMemo(() => {
-    return maybeSorted.map((row) => {
+  // Filter + flatten + component-ize the sorted rows.
+  let [headerRows, filteredRows]: [RowTuple<R>[], RowTuple<R>[]] = useMemo(() => {
```

Comment:
We used to have two steps, `allRows` was the react components and memo'd, and then we filtered against the flattened list.

When the change to nesting, both filtering and collapsing are best/easiest done pre-flattening, so now we have a single step that both flattens + filters + component-izes things. I'm pretty sure this is a net zero effect in terms of re-renders but we should probably have some like unit tests that use leaky state / global vars to count / assert against re-renders while filters/etc. change...